README.org
#+title: XMonad Config
#+author: Karim
#+property: header-args :tangle src/xmonad.hs

* Table of contents :toc:
- [[#imports][Imports]]
- [[#variables][Variables]]
- [[#workspaces][Workspaces]]
- [[#scratchpads][Scratchpads]]
- [[#log-hook][Log Hook]]
  - [[#status-bar][Status bar]]
- [[#startup-hook][Startup Hook]]
- [[#event-hook][Event Hook]]
- [[#manage-hook][Manage Hook]]
- [[#layout-hook][Layout Hook]]
- [[#keybindings][Keybindings]]
- [[#main][Main]]

* Imports
#+begin_src haskell

  -- Builtin
import Data.Map qualified as M
import Data.Monoid
import System.Exit

  -- Basic
import XMonad
import XMonad.StackSet qualified as W

  -- Layouts
import XMonad.Layout.MultiToggle
import XMonad.Layout.MultiToggle.Instances
import XMonad.Layout.ResizableThreeColumns
import XMonad.Layout.Spacing
import XMonad.Layout.Decoration
import XMonad.Layout.SimplestFloat
import XMonad.Layout.Grid
import XMonad.Layout.MultiToggle
import XMonad.Layout.MultiToggle.Instances
import XMonad.Layout.NoBorders
import XMonad.Layout.Renamed
import XMonad.Layout.ResizableTile
import XMonad.Layout.WindowArranger
import XMonad.Layout.BorderResize
import XMonad.Layout.PositionStoreFloat

  -- Actions
import XMonad.Actions.SpawnOn
import XMonad.Actions.MouseResize
import XMonad.Actions.CopyWindow
import XMonad.Actions.CycleWS

  -- Hooks
import XMonad.Hooks.InsertPosition
import XMonad.Hooks.Place
import XMonad.Hooks.ManageDocks
import XMonad.Hooks.StatusBar
import XMonad.Hooks.FadeWindows
import XMonad.Hooks.WorkspaceHistory
import XMonad.Hooks.WindowSwallowing
import XMonad.Hooks.SetWMName
import XMonad.Hooks.StatusBar
import XMonad.Hooks.StatusBar.PP
import XMonad.Hooks.ManageDocks
import XMonad.Hooks.EwmhDesktops
  -- Utils
import XMonad.Util.EZConfig
import XMonad.Util.NamedActions
import XMonad.Util.NamedScratchpad
import XMonad.Util.Hacks qualified as Hacks
import XMonad.Util.ClickableWorkspaces
import XMonad.Util.SpawnOnce

  -- MyLib
import Color.Theme
#+end_src


* Variables
#+begin_src haskell
myTerminal ::String
myTerminal = "alacritty"

-- Whether focus follows the mouse pointer.
myFocusFollowsMouse :: Bool
myFocusFollowsMouse = True

-- Whether clicking on a window to focus also passes the click to the window
myClickJustFocuses :: Bool
myClickJustFocuses = False

myModMask :: KeyMask
myModMask = mod4Mask

myNormalBorderColor :: String
myNormalBorderColor = "#5599cc"

myFocusedBorderColor :: String
myFocusedBorderColor = "#55ff99"

myBorderWidth :: Dimension
myBorderWidth = 2

defaultGapSize :: Integer
defaultGapSize = 10

fileManager :: String
fileManager = "thunar"

browser :: String
browser = "microsoft-edge-dev"

fileBrowser :: String
fileBrowser = "thunar"

pdfHistory :: String
pdfHistory = "~/script/pdf_history"

powerOpts :: String
powerOpts = "~/script/powerOptions"

doom_emacsclient :: String
doom_emacsclient = "~/script/run_emacs doom doom-emacs"

vanilla_emacsclient :: String
vanilla_emacsclient = "~/script/run_emacs vanilla vanilla-emacs"

dmenu_run :: String
dmenu_run = "~/Suckless/bin/dmenu_run_history"

taskManager :: String
taskManager = "btm"

changeThemeScript = "$XDG_CONFIG_HOME/xmonad/scripts/select-theme"


#+end_src

* Workspaces
Workspaces are the tags that appear in the status bar, each tag corresponds to virtual workspace.
#+begin_src haskell
myWorkspaces :: [String]
myWorkspaces = ["home","fecu","www","docs","dev","sys-mon"] --map show [1..9::Int]

#+end_src


* Scratchpads
#+begin_src haskell
scratchpads =
  [ -- run htop in xterm, find it by title, use default floating window placement
    NS "notes" "st -e nvim" (title =? "notes") defaultFloating,
    NS
      "stardict"
      "stardict"
      (className =? "Stardict")
      (customFloating $ W.RationalRect (1 / 6) (1 / 6) (2 / 3) (2 / 3))
  ]

#+end_src


* Log Hook
#+begin_src haskell

myLogHook :: X ()
myLogHook =
  workspaceHistoryHook

myFadeHook :: FadeHook
myFadeHook = composeAll []

#+end_src

** Status bar
This section defines a pretty printer (PP) for the status bar.

#+begin_src haskell
myPP :: PP
myPP =
  def
    { ppCurrent = xmobarColor (colorRed theme) "",
      ppUrgent = xmobarColor (colorGreen theme) (colorBPurple theme),
      ppLayout = xmobarFont 5 . xmobarColor (colorBPurple theme) "" ,
      ppSep = " ",
      ppWsSep = " ",
      ppExtras = [windowCount],
      ppTitle = xmobarColor (colorFore theme) "" . shorten 45,
      ppHidden = xmobarColor (colorCyan theme) "",
      ppHiddenNoWindows = xmobarColor (colorGrey theme) "",
      ppOrder = \(ws : l : t : ex) -> ws : ex ++ l : [t]
    }

windowCount :: X (Maybe String)
windowCount =
  Just
    . xmobarColor (colorGreen theme) ""
    . show
    . length
    . W.integrate'
    . W.stack
    . W.workspace
    . W.current
    . windowset
    <$> get

mySB :: StatusBarConfig
mySB =
  statusBarProp "xmobar" $
    copiesPP (xmobarFont 4 . xmobarColor (colorBlack theme) "") myPP >>= clickablePP

#+end_src


* Startup Hook
Startup hook
Perform an arbitrary action each time xmonad starts or is restarted
with mod-q.
`spawnOnce` is called only the first time xmonad is started
`spawn` is called every time xmonad is started
#+begin_src haskell

myStartupHook :: X ()
myStartupHook = do
  setWMName "LG3D"
  spawnOnce "sxhkd"
  spawnOnce "emacs --with-profile doom-emacs --daemon &"
  spawnOnce "emacs --with-profile vanilla-emacs --daemon &"

  spawnOnOnce "sys-mon" ("st -e btm")
  spawn "killall trayer"
  spawnOnce "nm-applet"
  spawnOnce "picom"

  spawn $
     "sleep 2 && trayer --edge top --align right --distance 10 --distancefrom right --distance 5 --distancefrom top \
      \--widthtype request --padding 6 --SetDockType true --SetPartialStrut false --expand true --transparent true --alpha 0 \
      \--tint "
        ++ trayerColor
        ++ " --height 30"


trayerColor = "0x" ++ tail (colorBack theme)

-- spawnOnOnce "2" browser
-- spawn "pcmanfm --desktop &"

#+end_src


* Event Hook
Defines a custom handler function for X Events. The function should
return (All True) if the default handler is to be run afterwards.

#+begin_src haskell
myEventHook :: Event -> X All
myEventHook =
  composeAll
    [ Hacks.windowedFullscreenFixEventHook,
      swallowEventHook (className =? "Alacritty" <||> className =? "Termite") (return True)
    ]
#+end_src


* Manage Hook
Execute arbitrary actions and WindowSet manipulations when managing
a new window. You can use this to, for example, always float a
particular program, or have a client always appear on a particular
workspace.

To find the property name associated with a program, use
> xprop | grep WM_CLASS
and click on the client you're interested in.

To match on the WM_NAME, you can use 'title' in the same way that
'className' and 'resource' are used below.

#+begin_src haskell
myManageHook :: Query (Endo WindowSet)
myManageHook =
  composeAll
    [ manageSpawn,
      insertPosition Below Newer,
      placeHook $ withGaps (16, 16, 16, 16) (smart (0.5, 0.5)), -- simpleSmart -- (smart (0.5,0.5))
      namedScratchpadManageHook scratchpads,
      className =? "jetbrains-idea-ce" --> doFloat,
      className =? "dialog" --> doFloat,
      className =? "download" --> doFloat,
      className =? "notification" --> doFloat,
      className =? "Xmessage" --> doFloat
    ]

#+end_src


* Layout Hook
The following layouts are supported
|----------------+--------------------+------------------------------------------------------------------+-----------|
| layout         | name in status bar | description                                                      | resizable |
|----------------+--------------------+------------------------------------------------------------------+-----------|
| resizableTiled | tall               | default layout, the screen is divided into master and stack.     | yes       |
| threeColMid    | three col mid      | divide the screen into three cols, the middle one is the master. | yes       |
| threeCol       | three col          | divide the screen into three cols, the leftmost is the master.   | yes       |
| grid           | grid               | divide the screen into equally sized rectangles.                 | no        |
| full           | monocle            | single window spanning the whole screen.                         | -         |
| myFloat        | flaot              | floating window layout, resizable using mouse or keyboard.       | yes       |
|----------------+--------------------+------------------------------------------------------------------+-----------|

#+begin_src haskell
mySpacing :: Integer -> Integer -> l a -> ModifiedLayout Spacing l a
mySpacing i j = spacingRaw False (Border i i i i) True (Border j j j j) True

resizableTiled = renamed [Replace "tall"]
               $ mySpacing defaultGapSize defaultGapSize
               $ ResizableTall 1 (3 / 100) (1 / 2) []

threeColMid = renamed [Replace "threeColMid"]
            $ mySpacing defaultGapSize defaultGapSize
            $ ResizableThreeColMid 1 (3 / 100) (1 / 2) []

threeCol = renamed [Replace "threeCol"]
            $ mySpacing defaultGapSize defaultGapSize
            $ ResizableThreeCol 1 (3 / 100) (1 / 2) []

grid = renamed [Replace "grid"] $ mySpacing defaultGapSize defaultGapSize Grid

full = renamed [Replace "monocle"] $ mySpacing defaultGapSize defaultGapSize Full

myFloat = renamed [Replace "float"]
        . mouseResize
        . borderResize
        . windowArrangeAll
        $ simplestFloat

myLayout = avoidStruts
         . smartBorders
         . mkToggle (NOBORDERS ?? FULL ?? EOT)
         . mkToggle (single MIRROR) $ lll -- . avoidStruts lll
  where
    lll =
            resizableTiled
        ||| threeCol
        ||| threeColMid
        ||| grid
        ||| full
        ||| myFloat

data OnlyTiled = OnlyTiled
  deriving (Read, Show)

instance SetsAmbiguous OnlyTiled where
  hiddens _ _ _ mst wrs = filter (`elem` W.integrate' mst) $ map fst wrs

#+end_src


* Keybindings
#+begin_src haskell

myKeys conf =
    subTitle "XMonad Controls"
    [ ("M-q"          , addName "Restart XMonad"                $ sbCleanupHook mySB >> spawn "xmonad --restart")
    , ("M-r"          , addName "Recompile XMonad"              $ spawn "xmonad --recompile")
    , ("M-S-c"        , addName "Kill the focused application"  $ kill1)
    , ("M-S-q"        , addName "Exit XMonad"                   $ io exitSuccess)
    , ("M-S-r"        , addName "Refresh XMonad"                $ refresh)
    , ("M-S-t"        , addName "Change color theme"            $ spawn changeThemeScript)
    , ("M-S-b"        , addName "Show/Hide status bar"          $ sendMessage ToggleStruts)
    ]
    ++
    subTitle "Dmenu Scripts"
    [ ("M-x"          , addName "Poweroff prompt "              $ spawn powerOpts)
    , ("M-p"          , addName "Dmenu app launcher"            $ spawn dmenu_run)
    , ("M-o"          , addName "Navigate your pdf history"     $ spawn pdfHistory)
    ]
    ++
    subTitle "Applications"
    [ ("M-S-<Return>" , addName ("Open a new terminal ("++myTerminal++")") $ spawn (terminal conf))
    , ("M-d"          , addName "Launch Doom Emacs"                        $ spawn doom_emacsclient)
    , ("M-v"          , addName "Launch vanilla Emacs"                     $ spawn vanilla_emacsclient)
    ]
    ++
    subTitle "Layout Controls"
    [ ("M-S-<Tab>"    , addName "Reset the window layout"             $ setLayout $ layoutHook conf)
    , ("M-<Tab>"      , addName "Next layout"                         $ sendMessage NextLayout)
    , ("M-S-m"        , addName "Rotate layout by 90 degrees"         $ sendMessage $ Toggle MIRROR)
    , ("M-t s"        , addName "Toggle gaps"                         $ toggleSpaces)
    , ("M-t b"        , addName "Toggle borders"                      $ sendMessage $ Toggle NOBORDERS)
    , ("M-t t"        , addName "Tile a floating window"              $ withFocused $ windows . W.sink)
    , ("M-,"          , addName "Increase windows in the master pane" $ sendMessage (IncMasterN 1))
    , ("M-."          , addName "Decrease windows in the master pane" $ sendMessage (IncMasterN (-1)))
    , ("M-S-n"        , addName "Open a scratchpad"                   $ namedScratchpadAction scratchpads "notes")
    ]
    ++
    subTitle "Window/Workspace Controls"
    [ ("M-C-a"        , addName "Copy the focused window to all workspaces" $ windows copyToAll)
    , ("M-S-a"        , addName "Kill all copies of the focused window"     $ killAllOtherCopies)
    , ("M-j"          , addName "Focus the next window"                     $ windows W.focusDown)
    , ("M-k"          , addName "Focus the previous window"                 $ windows W.focusUp)
    , ("M-<Return>"   , addName "Swap the focused window with the master window"   $ windows W.swapMaster)
    , ("M-S-j"        , addName "Swap the focused window with the next window"     $ windows W.swapDown)
    , ("M-S-k"        , addName "Swap the focused window with the previous window" $ windows W.swapUp)
    , ("M-h"          , addName "Shrink window"       $ sendMessage Shrink)
    , ("M-l"          , addName "Expand window"       $ sendMessage Expand)
    , ("M-S-l"        , addName "MirrorShrink window" $ sendMessage MirrorShrink)
    , ("M-S-h"        , addName "MirrorExpand window" $ sendMessage MirrorExpand)
    , ("M-S-<Right>"  , addName "Shift window to next workspace"             $ shiftToNext)
    , ("M-S-<Left>"   , addName "Shift window to prev workspace"             $ shiftToPrev)
    , ("M-C-<Right>"  , addName "Shift window to next workspace, then goto"  $ shiftToNext >> nextWS)
    , ("M-C-<Left>"   , addName "Shift window to prev workspace, then goto"  $ shiftToPrev >> prevWS)
    ]
    ++
    subTitle "Floating Layouts Controls"
    [ ("M-M1-<Left>"  , addName "Move window left by 10 pixels"      $ sendMessage (MoveLeft 10))
    , ("M-M1-<Right>" , addName "Move window right by 10 pixels"     $ sendMessage (MoveRight 10))
    , ("M-M1-<Down>"  , addName "Move window down by 10 pixels"      $ sendMessage (MoveDown 10))
    , ("M-M1-<Up>"    , addName "Move window up by 10 pixels"        $ sendMessage (MoveUp 10))
    , ("M1-C-<Left>"  , addName "Expand the left edge by 5 pixels"   $ sendMessage (IncreaseLeft 5))
    , ("M1-C-<Right>" , addName "Expand the right edge by 5 pixels"  $ sendMessage (IncreaseRight 5))
    , ("M1-C-<Down>"  , addName "Expand the bottom edge by 5 pixels" $ sendMessage (IncreaseDown 5))
    , ("M1-C-<Up>"    , addName "Expand the top edge by 5 pixels"    $ sendMessage (IncreaseUp 5))

CHANGELOG.md
# Changelog for `xmonad-config`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


app/xmonad.hs
{-# LANGUAGE FlexibleContexts, PartialTypeSignatures #-}
{-# OPTIONS_GHC -Wno-deprecations #-}
{-# OPTIONS_GHC -Wno-missing-signatures #-}
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Use camelCase" #-}
{-# OPTIONS_GHC -Wno-unused-top-binds #-}
{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
{-# LANGUAGE FlexibleInstances, MultiParamTypeClasses #-}
---------------------------------------------------------------
import           XMonad.Actions.CopyWindow
import           XMonad.Actions.CycleWS
import           XMonad.Actions.SpawnOn
import           XMonad.Actions.Minimize
import           XMonad.Hooks.EwmhDesktops
import           XMonad.Hooks.FadeWindows
import           XMonad.Hooks.ManageDocks
import           XMonad.Hooks.StatusBar
import           XMonad.Hooks.SetWMName
import           XMonad.Hooks.Minimize
import           XMonad.Hooks.Place
import           XMonad.Actions.TreeSelect
import           XMonad.Hooks.WorkspaceHistory
import           XMonad.Hooks.InsertPosition
import           XMonad.Hooks.WindowSwallowing
---------------------------------------------------------------
---------------------------------------------------------------
import           XMonad.Layout.MagicFocus
import           XMonad.Layout.NoBorders
import           XMonad.Layout.Decoration
import           XMonad.Layout.Grid
import           XMonad.Layout.MultiToggle
import           XMonad.Layout.MultiToggle.Instances
import           XMonad.Layout.ResizableThreeColumns
import           XMonad.Layout.ResizableTile
import           XMonad.Layout.Spacing
import           XMonad.Layout.WindowArranger
import           XMonad.Layout.Maximize
import           XMonad.Layout.BoringWindows
----------------------------------------------------------------
----------------------------------------------------------------
import qualified Data.Map as M
import           Data.Tree
import           Data.Monoid
import           Data.List
import           System.Exit
import           XMonad
import qualified XMonad.StackSet as W
import           XMonad.Util.SpawnOnce
import qualified XMonad.Util.Hacks as Hacks
import           XMonad.Util.NamedScratchpad
-----------------------------------------------------------------
-----------------------------------------------------------------
import           Bar
import qualified FloatingLayouts as FL
-----------------------------------------------------------------
-----------------------------------------------------------------
-- main
main :: IO ()
main = xmonad . withSB mySB . docks . ewmhFullscreen . ewmh $ defaults

-- A structure containing your configuration settings, overriding
-- fields in the default config. Any you don't override, will
-- use the defaults defined in xmonad/XMonad/Config.hs
--
defaults {-fullScreenRef-} = def   -- simple stuff
  { terminal = myTerminal
  , focusFollowsMouse = myFocusFollowsMouse
  , clickJustFocuses = myClickJustFocuses
  , borderWidth = myBorderWidth
  , modMask = myModMask
  , workspaces =  myWorkspaces
  , normalBorderColor = myNormalBorderColor
  , focusedBorderColor = myFocusedBorderColor
  , keys = myKeys -- fullScreenRef
  , mouseBindings = myMouseBindings
  , layoutHook = myLayout
  , manageHook = myManageHook
  , handleEventHook = myEventHook
  , logHook = myLogHook
  , startupHook = myStartupHook
  }

------------------------------------------------------------------------
-- Key bindings. Add, modify or remove key bindings here.
--
myKeys    {-IORef Bool ->-} :: XConfig Layout -> M.Map (KeyMask, KeySym) (X ())
myKeys {-fullScreenRef-} conf@XConfig { modMask = modm } = M.fromList
  $ [ ((modm .|. shiftMask    , xK_Return)                , spawn $ terminal conf)
    , ((modm                  , xK_x)                     , spawn powerOpts)
    , ((modm                  , xK_p)                     , spawn dmenu_run)
    , ((modm .|. shiftMask    , xK_space)                 , setLayout $ layoutHook conf)
    , ((modm                  , xK_n)                     , refresh)
    , ((modm .|. controlMask  , xK_a)                     , windows copyToAll)
    , ((modm .|. shiftMask    , xK_a)                     , killAllOtherCopies)
    , ((modm                  , xK_Tab)                   , {-windows W.-}focusDown)
    , ((modm                  , xK_j)                     , {-windows W.-}focusDown)
    , ((modm                  , xK_k)                     , {-windows W.-}focusUp)
    , ((modm                  , xK_m)                     , windows W.focusMaster)
    , ((modm                  , xK_Return)                , windows W.swapMaster)
    , ((modm .|. shiftMask    , xK_j)                     , windows W.swapDown)
    , ((modm .|. shiftMask    , xK_k)                     , windows W.swapUp)
    , ((modm .|. shiftMask    , xK_c)                     , kill1)
    , ((modm                  , xK_d)                     , spawn doom_emacsclient)
    , ((modm                  , xK_v)                     , spawn vanilla_emacsclient)
    , ((modm                  , xK_o)                     , spawn pdfHistory)
    , ((modm                  , xK_space)                 , sendMessage NextLayout)
    , ((modm                  , xK_r)                     , sendMessage $ Toggle MIRROR)
    , ((modm                  , xK_h)                     , sendMessage Shrink)
    , ((modm                  , xK_l)                     , sendMessage Expand)
    , ((modm .|. shiftMask    , xK_l)                     , sendMessage MirrorShrink)
    , ((modm .|. shiftMask    , xK_h)                     , sendMessage MirrorExpand)
    , ((modm                  , xK_b)                     , sendMessage ToggleStruts)
    , ((modm                  , xK_s)                     , toggleSpaces) -- >> withAll toggleBorder ) 
    , ((modm                  , xK_t)                     , withFocused $ windows . W.sink)
    , ((modm                  , xK_comma)                 , sendMessage (IncMasterN 1))
    , ((modm                  , xK_period)                , sendMessage (IncMasterN (-1)))
    , ((modm .|. shiftMask    , xK_q)                     , io exitSuccess)
    , ((modm                  , xK_q)                     , sbCleanupHook mySB >> spawn xmonadRestartComm) -- sbCleanupHook mySB >>
    -- , ((modm                  , xK_q)                     , spawn "xmonad --recompile" >> restart "xmonad" True)
      --------------------------------------------------------------------------------------------------
    -- , ((modm                  , xK_s)                     , submap . M.fromList $ spaceControl modm)
    -- , ((modm                  , xK_a)                     , submap . M.fromList $ applications modm)
    , ((modm .|. shiftMask    , xK_s    )                 , sendMessage  Arrange         )
    , ((modm .|. controlMask  , xK_s    )                 , sendMessage  DeArrange       )
     ------------------------------------------------------------------------------------------------
    , ((modm .|. shiftMask    , xK_n)                     , namedScratchpadAction scratchpads "notes")
    , ((modm                  , xK_Right)                 , nextWS)
    , ((modm                  , xK_Left)                  , prevWS)
    , ((modm .|. shiftMask    , xK_Right)                 , shiftToNext)
    , ((modm .|. shiftMask    , xK_Left)                  , shiftToPrev)
    , ((modm .|. controlMask  , xK_Right)                 , shiftToNext >> nextWS)
    , ((modm .|. controlMask  , xK_Left)                  , shiftToPrev >> prevWS)
    , ((modm                  , xK_backslash)             , withFocused (sendMessage . maximizeRestore))
    , ((modm                  , xK_i)                     , withFocused minimizeWindow      )
    , ((modm .|. shiftMask    , xK_i)                     , withLastMinimized maximizeWindowAndFocus)

      ------------------------------------------------------------------------------------------------
  ] ++
  [ ((m .|. modm              , k)                        , windows $ f i)
  | (i, k) <- zip (workspaces conf) [xK_1 .. xK_9]
  , (f, m) <- [(W.greedyView, 0)
  , (W.shift, shiftMask), (copy, controlMask)]]




applications modm = [ ((modm, xK_d), spawn doom_emacsclient)
                    , ((modm, xK_f), spawn "alacritty -e nnn")
                    , ((modm, xK_b), spawn browser)
                    , ((modm, xK_Alt_L), spawn fileBrowser)]

spaceControl modm = [ ((modm, xK_KP_Add), incScreenWindowSpacing 5)
                    , ((modm, xK_KP_Subtract), decScreenWindowSpacing 5)
                    , ((modm, xK_KP_Insert), setScreenWindowSpacing 5)]

toggleSpaces :: X ()
toggleSpaces = toggleScreenSpacingEnabled >> toggleWindowSpacingEnabled

------------------------------------------------------------------------
-- Mouse bindings: default actions bound to mouse events
--
myMouseBindings :: XConfig l -> M.Map (KeyMask, Button) (Window -> X ())
myMouseBindings XConfig { XMonad.modMask = modm } = M.fromList
  -- mod-button1, Set the window to floating mode and move by dragging
  [ ( (modm, button1)
    , \w -> focus w >> mouseMoveWindow w >> windows W.shiftMaster)
    -- mod-button2, Raise the window to the top of the stack
  , ((modm .|. controlMask, button1), \w -> focus w >> windows W.shiftMaster)
    -- mod-button3, Set the window to floating mode and resize by dragging
  , ( (modm .|. shiftMask, button1)
    , \w -> focus w >> mouseResizeWindow w >> windows W.shiftMaster)]

      -- you may also bind events to the mouse scroll wheel (button4 and button5)
------------------------------------------------------------------------
-- Layouts:
-- You can specify and transform your layouts by modifying these values.
-- If you change layout bindings be sure to use 'mod-shift-space' after
-- restarting (with 'mod-q') to reset your layout state to the new
-- defaults, as xmonad preserves your old layout settings by default.
--
-- The available layouts.  Note that each layout is separated by |||,
-- which denotes layout choice.
--


spacingRawValue :: Integer -> l a -> ModifiedLayout Spacing l a
spacingRawValue i = spacingRaw False (Border i i i i) True (Border i i i i) True

resizableTiled = mkToggle (single MIRROR) 
               $ spacingRawValue 5 
               $ ResizableTall 1 (3 / 100) (1 / 2) []


threeColMid = mkToggle (single MIRROR) 
            $ spacingRawValue  5 
            $ ResizableThreeColMid 1 (3 / 100) (1 / 2) []

grid = spacingRawValue 5 Grid

full = Full

-- Width of the window border in pixels.
--
myBorderWidth :: Dimension
myBorderWidth = 2



myLayout = avoidStruts $ boringWindows lll -- $ mouseResize $ windowArrange   lll
  where
    lll = 
      resizableTiled     ||| 
      threeColMid        ||| 
      grid               |||
      full               |||   
      FL.floatLayout 
      
   
   
data OnlyTiled = OnlyTiled
  deriving (Read, Show)

instance SetsAmbiguous OnlyTiled where
  hiddens _ _ _ mst wrs = filter (`elem` W.integrate' mst) $ map fst wrs


------------------------------------------------------------------------
-- Window rules:
-- Execute arbitrary actions and WindowSet manipulations when managing
-- a new window. You can use this to, for example, always float a
-- particular program, or have a client always appear on a particular
-- workspace.
--
-- To find the property name associated with a program, use
-- > xprop | grep WM_CLASS
-- and click on the client you're interested in.
--
-- To match on the WM_NAME, you can use 'title' in the same way that
-- 'className' and 'resource' are used below.
--
myManageHook :: Query (Endo WindowSet)
myManageHook = composeAll [
         manageSpawn
        , insertPosition Below Newer
        , placeHook $ withGaps (16,16,16,16) (smart (0.5,0.5)) -- simpleSmart -- (smart (0.5,0.5))
        , namedScratchpadManageHook scratchpads
        , className =? "jetbrains-idea-ce" --> doFloat 
        , className =? "dialog"          --> doFloat
        , className =? "download"        --> doFloat
        , className =? "notification"    --> doFloat
       ] 

  -- , liftX current_is_floating --> hasBorder True -- Borders around floating windows
 --  , className =? "VScodium" --> doFloat
  -- , className =? "Gimp" --> doFloat
  -- , resource =? "desktop_window" --> doIgnore
  -- , resource =? "kdesktop" --> doIgnore

current_is_floating :: X Bool
current_is_floating = do
  wins <- gets windowset
  return
    $ case W.peek wins of
      Just w  -> M.member w (W.floating wins)
      Nothing -> False

------------------------------------------------------------------------
-- Event handling
-- * EwmhDesktops users should change this to ewmhDesktopsEventHook
--
-- Defines a custom handler function for X Events. The function should
-- return (All True) if the default handler is to be run afterwards. To
-- combine event hooks use mappend or mconcat from Data.Monoid.
--
myEventHook :: Event -> X All
myEventHook = composeAll [ 
     Hacks.windowedFullscreenFixEventHook 
    , fadeWindowsEventHook 
    , swallowEventHook (className =? "Alacritty" <||> className =? "Termite") (return True)
    , minimizeEventHook
    , followOnlyIf (fmap not isLayoutFloat)
    ]

isLayoutFloat :: X Bool
isLayoutFloat = fmap (isSuffixOf "Floating") $ gets (description . W.layout . W.workspace . W.current . windowset)

    --  <+> toggleableFullscreen fullScreenRef
------------------------------------------------------------------------
-- Status bars and logging
-- Perform an arbitrary action on each internal state change or X event.
-- See the 'XMonad.Hooks.DynamicLog' extension for examples.
--
myLogHook :: X ()
myLogHook = fadeWindowsLogHook myFadeHook <+> workspaceHistoryHook

myFadeHook :: FadeHook
myFadeHook = composeAll [opaque, className =? "dmenu" --> transparency 0.5]

------------------------------------------------------------------------
-- Startup hook
-- Perform an arbitrary action each time xmonad starts or is restarted
-- with mod-q.  Used by, e.g., XMonad.Layout.PerWorkspace to initialize
-- per-workspace layout choices.
--
-- By default, do nothing.
myStartupHook :: X ()
myStartupHook = do
  setWMName "LG3D"
  spawnOnce "sxhkd"
  spawnOnOnce "sys-mon" "st -e btm"

-- spawnOnOnce "2" browser
-- spawn "pcmanfm --desktop &"
------------------------------------------------------------------------
-- The preferred terminal program, which is used in a binding below and by
-- certain contrib modules.
--
myTerminal :: String
myTerminal = "alacritty"

-- Whether focus follows the mouse pointer.
myFocusFollowsMouse :: Bool
myFocusFollowsMouse = True

-- Whether clicking on a window to focus also passes the click to the window
myClickJustFocuses :: Bool
myClickJustFocuses = False

myModMask :: KeyMask
myModMask = mod4Mask

dwmBlue :: String
dwmBlue = "#2b4f98"

fileManager :: String
fileManager = "thunar"

xmonadRestartComm :: String
xmonadRestartComm = "killall xmobar; xmonad --recompile; xmonad --restart"

xmobarConfigPath :: String
xmobarConfigPath = "xmobar "

rofiCommand :: String
rofiCommand = "/home/karim/.config/rofi/launchers/misc/launcher.sh"

browser :: String
browser = "microsoft-edge-dev"

fileBrowser :: String
fileBrowser = "thunar"

myBin :: String
myBin = "~/bin/"

pdfHistory :: String
pdfHistory = "~/script/pdf_history"

powerOpts :: String
powerOpts = "~/script/powerOptions"

doom_emacsclient :: String
doom_emacsclient = "~/script/run_emacs doom-daemon default"

vanilla_emacsclient::String
vanilla_emacsclient = "~/script/run_emacs vanilla-daemon vanilla"

dmenu_run :: String
dmenu_run = "~/Suckless/bin/dmenu_run_history"

dmenu_browse :: String
dmenu_browse = myBin ++ "dbrowse"

taskManager :: String
taskManager = "btm"


------------------------------------------------------------------------
-- The default number of workspaces (virtual screens) and their names.
-- By default we use numeric strings, but any string may be used as a
-- workspace name. The number of workspaces is determined by the length
-- of this list.
--
-- A tagging example:
--
-- myWorkspaces = ["web", "irc", "code" ] ++ map show [4..9]
--
myWorkspaces :: [String]
myWorkspaces = ["home","dev","web","docs","fecu","sys-mon"] --map show [1..9::Int] 
-- myWorkspaces :: Forest String
-- myWorkspaces =
--   [ Node
--       "doc"       -- for everyday activity's
--       [ Node "1" []   --  with 4 extra sub-workspaces, for even more activity's
--       , Node "2" []
--       , Node "3" []
--       , Node "4" []
--   , Node
--       "dev" -- for all your programming needs
--       [ Node "haskell" []
--       , Node "lean" [] -- documentation
--       ]
--   , Node "taskmanager" []    ]

wsMenuWidth = 100
wsMenuHight = 18

wsMenuConf :: TSConfig a
wsMenuConf =
  def { ts_background = 0x00000000
      , ts_font = "xft:Noto Sans:size=14"
      , ts_node = (0xffffffff, 0xff004466)
      , ts_nodealt = (0xffffffff, 0xff004466)
      , ts_highlight = (0xff0000ff, 0xff50d0db)
      , ts_extra = 0x00000000
      , ts_node_width = 2*wsMenuWidth
      , ts_node_height = 2*wsMenuHight
      , ts_originX = 0
      , ts_originY = 0
      , ts_indent = 80
      }

powerMenuWidth = 100
powerMenuHight = 18

powerMenuConf :: TSConfig a
powerMenuConf =
  def { ts_background = 0x00000000
      , ts_font = "xft:Noto Sans:size=14"
      , ts_node = (0xffffffff, 0xff004466)
      , ts_nodealt = (0xffffffff, 0xff004466)
      , ts_highlight = (0xff0000ff, 0xff50d0db)
      , ts_extra = 0x00000000
      , ts_node_width = 2 * powerMenuWidth
      , ts_node_height = 2*powerMenuHight
      , ts_originX = 960 - powerMenuWidth
      , ts_originY = 540-powerMenuHight
      , ts_indent = 80
      }

-- Border colors for unfocused and focused windows, respectively.
--
myNormalBorderColor :: String
myNormalBorderColor = "#5599cc"

myFocusedBorderColor :: String
myFocusedBorderColor = "#55ff99" 

powerMenu = treeselectAction
  powerMenuConf
  [ Node (TSNode "Shutdown" "" (spawn "poweroff")) []
  , Node (TSNode "Hibernate" "" (spawn "systemctl hibernate")) []
  , Node (TSNode "Reboot" "" (spawn "reboot")) []
  , Node (TSNode "Logout" "" (io exitSuccess)) []
  , Node (TSNode "Cancel" "" (return ())) []]


scratchpads= [
-- run htop in xterm, find it by title, use default floating window placement
    NS "notes" "st -e nvim" (title =? "notes") defaultFloating ,

    NS "stardict" "stardict" (className =? "Stardict")
        (customFloating $ W.RationalRect (1/6) (1/6) (2/3) (2/3)) 

    ] 

